lib/fork.js

```diff
@@ -1,132 +1,107 @@
-		cwd: opts.projectDir,
-		silent: true,
-		env,
-		execArgv: execArgv || process.execArgv
-	});
-
-	subprocess.stdout.on('data', chunk => {
-		emitStateChange({type: 'worker-stdout', chunk});
-	});
-
-	subprocess.stderr.on('data', chunk => {
-		emitStateChange({type: 'worker-stderr', chunk});
-	});
-
-	let forcedExit = false;
-	const send = evt => {
-		if (subprocess.connected && !finished && !forcedExit) {
-			subprocess.send({ava: evt});
-		}
-	};
-
-	const promise = new Promise(resolve => {
-		const finish = () => {
-			finished = true;
-			resolve();
-		};
-
-		subprocess.on('message', message => {
-			if (!message.ava) {
-				return;
-			}
-
-			if (message.ava.type === 'ping') {
-				send({type: 'pong'});
-			} else {
-				emitStateChange(message.ava);
-			}
-		});
+const subprocessPath = require.resolve('./worker/subprocess.js');
 
-		subprocess.on('error', err => {
-			emitStateChange({type: 'worker-failed', err});
-			finish();
-		});
+const failTypes = {
+	'hook-failed': 1,
+	'test-failed': 1,
+	'worker-failed': 1
+};
 
-		subprocess.on('exit', (code, signal) => {
-			if (forcedExit) {
-				emitStateChange({type: 'worker-finished', forcedExit});
-			} else if (code > 0) {
-				emitStateChange({type: 'worker-failed', nonZeroExitCode: code});
-			} else if (code === null && signal) {
-				emitStateChange({type: 'worker-failed', signal});
-			} else {
-				emitStateChange({type: 'worker-finished', forcedExit});
-			}
-
-			finish();
+// Create a forked subprocess to run `files` under
+class Fork {
+	/* eslint-disable-next-line max-params */
+	constructor(files, runStatus, env, opts, execArgv) {
+		this.runStatus = runStatus;
+		this.options = Object.assign({
+			baseDir: process.cwd(),
+			tty: {
+				stderr: process.stderr.isTTY ? describeTTY(process.stderr) : false,
+				stdout: process.stdout.isTTY ? describeTTY(process.stdout) : false
+			},
+			files
+		}, opts);
+
+		this.projectDir = opts.api.options.projectDir;
+		this.subprocess = null;
+		this.execArgv = execArgv || process.execArgv;
+		this.env = env;
+		this.args = [
+			JSON.stringify(this.options),
+			opts.color ? '--color' : '--no-color'
+		]
+			.concat(opts.api.options.workerArgv || []);
+	}
+
+	run() {
+		return new Promise((resolve, reject) => {
+			const {options, runStatus} = this;
+			const {failFast} = options;
+
+			this.subprocess = childProcess.fork(subprocessPath, this.args, {
+				cwd: this.projectDir,
+				silent: true,
+				env: this.env,
+				execArgv: this.execArgv
+			});
+
+			// Handle stdout and stderr
+			this.subprocess.stdout.on('data', chunk => {
+				runStatus.emitStateChange({type: 'worker-stdout', chunk});
+			});
+
+			this.subprocess.stderr.on('data', chunk => {
+				runStatus.emitStateChange({type: 'worker-stderr', chunk});
+			});
+
+			// Listen for messages
+			this.subprocess.on('message', msg => {
+				if (msg.forkMsg === 'new-file') {
+					runStatus.observeWorker({onStateChange: () => {}}, msg.file);
```

Comment:
Looks like `run-status` needs updating?

Reply:
After re-working the branch, I don't think `run-status` need updating anymore. Do you think it still does?